Release proposal: v3.0.0

rvagg opened the pull request:
I'm tackling this 3.0.0 problem from a different angle, I'm sick of producing "RC" builds from a `next` branch that is so far behind `master`, meaning they are nothing like "release candidates" and catching up is next to impossible.

This branch, on nodejs/io.js so others can contribute, is the current `master` with changes from `next` cherry-picked in and I'm just going to rebase to keep it up to date and it's mergable if we decide to go that way (I'm leaning that way currently fwiw).

I made a [branch-diff](https://github.com/rvagg/iojs-tools/tree/master/branch-diff) tool that's like doing `git log b1..b2` but not as strict and uses our `PR-URL` metadata to help with matches. This was mainly to help identify what to cherry-pick. It also prints out changelog-maker output so I can also do this:

* [[`78b4a07085`](https://github.com/nodejs/io.js/commit/78b4a07085)] - **node-gyp**: detect RC build with x.y.z-rc.n format (Rod Vagg) [#2171](https://github.com/nodejs/io.js/pull/2171)
* [[`62eeeb0f45`](https://github.com/nodejs/io.js/commit/62eeeb0f45)] - **src**: introduce process.release object (Rod Vagg) [#2154](https://github.com/nodejs/io.js/pull/2154)
* [[`8be6060020`](https://github.com/nodejs/io.js/commit/8be6060020)] - **http**: use official IANA Status Codes (jomo) [#1470](https://github.com/nodejs/io.js/pull/1470)
* [[`4b844e68f7`](https://github.com/nodejs/io.js/commit/4b844e68f7)] - **node**: do not override `message`/`stack` of error (Fedor Indutny) [#2108](https://github.com/nodejs/io.js/pull/2108)
* [[`4a4775e821`](https://github.com/nodejs/io.js/commit/4a4775e821)] - **src**: increment NODE_MODULE_VERSION to 45 (Rod Vagg) [#2096](https://github.com/nodejs/io.js/pull/2096)
* [[`d067cdbbfd`](https://github.com/nodejs/io.js/commit/d067cdbbfd)] - **deps**: upgrade v8 to 4.4.63.12 (Ben Noordhuis) [#2092](https://github.com/nodejs/io.js/pull/2092)
* [[`d8355b7d6f`](https://github.com/nodejs/io.js/commit/d8355b7d6f)] - Working on v3.0.0 (Rod Vagg) 
* [[`11d9b332de`](https://github.com/nodejs/io.js/commit/11d9b332de)] - **node-gyp**: download header tarball for compile (Rod Vagg) [#2066](https://github.com/nodejs/io.js/pull/2066)
* [[`9e41d1de3b`](https://github.com/nodejs/io.js/commit/9e41d1de3b)] - **node-gyp**: make aware of nightly, next-nightly & rc (Rod Vagg) 
* [[`8346c5029a`](https://github.com/nodejs/io.js/commit/8346c5029a)] - **buffer**: rename internal/buffer_new.js to buffer.js (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022)
* [[`42c8b8030e`](https://github.com/nodejs/io.js/commit/42c8b8030e)] - **test**: remove two obsolete pummel tests (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022)
* [[`13446092af`](https://github.com/nodejs/io.js/commit/13446092af)] - **test**: don't use arguments.callee (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022)
* [[`9f6f200c87`](https://github.com/nodejs/io.js/commit/9f6f200c87)] - **deps**: update v8 to 4.4.63.9 (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022)
* [[`717c634a60`](https://github.com/nodejs/io.js/commit/717c634a60)] - **buffer**: allow ArrayBuffer as Buffer argument (Trevor Norris) [#2002](https://github.com/nodejs/io.js/pull/2002)
* [[`9743d1f53d`](https://github.com/nodejs/io.js/commit/9743d1f53d)] - **buffer**: minor cleanup from rebase (Trevor Norris) [#2003](https://github.com/nodejs/io.js/pull/2003)
* [[`535b6f34e9`](https://github.com/nodejs/io.js/commit/535b6f34e9)] - **buffer**: fix usage of kMaxLength (Trevor Norris) [#2003](https://github.com/nodejs/io.js/pull/2003)
* [[`e0d8ca18d5`](https://github.com/nodejs/io.js/commit/e0d8ca18d5)] - **cluster**: do not unconditionally set --debug-port (cjihrig) [#1949](https://github.com/nodejs/io.js/pull/1949)
* [[`12b15aaa7c`](https://github.com/nodejs/io.js/commit/12b15aaa7c)] - **buffer**: fix case of one buffer passed to concat (Sakthipriyan Vairamani) [#1937](https://github.com/nodejs/io.js/pull/1937)
* [[`ce1acd5337`](https://github.com/nodejs/io.js/commit/ce1acd5337)] - **http_server**: `prefinish` vs `finish` (Fedor Indutny) [#1411](https://github.com/nodejs/io.js/pull/1411)
* [[`acac161642`](https://github.com/nodejs/io.js/commit/acac161642)] - **dgram**: make send cb act as "error" event handler (Matteo Collina) [#1796](https://github.com/nodejs/io.js/pull/1796)
* [[`6b3f046f77`](https://github.com/nodejs/io.js/commit/6b3f046f77)] - **dgram**: make send cb act as "error" event handler (Chris Dickinson) [#1796](https://github.com/nodejs/io.js/pull/1796)
* [[`81f12be662`](https://github.com/nodejs/io.js/commit/81f12be662)] - **vm**: fix symbol access (Domenic Denicola) [#1773](https://github.com/nodejs/io.js/pull/1773)
* [[`222c8a53cf`](https://github.com/nodejs/io.js/commit/222c8a53cf)] - **vm**: remove unnecessary access checks (Domenic Denicola) [#1773](https://github.com/nodejs/io.js/pull/1773)
* [[`37a8a54fb2`](https://github.com/nodejs/io.js/commit/37a8a54fb2)] - **vm**: fix property descriptors of sandbox properties (Domenic Denicola) [#1773](https://github.com/nodejs/io.js/pull/1773)
* [[`d4ee3901f7`](https://github.com/nodejs/io.js/commit/d4ee3901f7)] - **crypto**: remove kMaxLength on randomBytes() (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`54e9a0b56d`](https://github.com/nodejs/io.js/commit/54e9a0b56d)] - **buffer**: make additional changes to native API (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`d6fb71086b`](https://github.com/nodejs/io.js/commit/d6fb71086b)] - **buffer**: switch API to return MaybeLocal<T> (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`8931e5b340`](https://github.com/nodejs/io.js/commit/8931e5b340)] - **buffer**: switch to using Maybe<T> API (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`9246155967`](https://github.com/nodejs/io.js/commit/9246155967)] - **buffer**: finish implementing FreeCallback (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`c735f738ef`](https://github.com/nodejs/io.js/commit/c735f738ef)] - **buffer**: implement Uint8Array backed Buffer (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`bec924f0ff`](https://github.com/nodejs/io.js/commit/bec924f0ff)] - **buffer**: allow ARGS_THIS to accept a name (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825)
* [[`8335ece6d8`](https://github.com/nodejs/io.js/commit/8335ece6d8)] - **deps**: backport 7b24219346 from v8 upstream (Rod Vagg) [#1805](https://github.com/nodejs/io.js/pull/1805)
* [[`a2ac312e56`](https://github.com/nodejs/io.js/commit/a2ac312e56)] - **net**: do not set V4MAPPED on FreeBSD (Julien Gilli) [iojs/io.js#1555](https://github.com/iojs/io.js/pull/1555)
* [[`0b122d09c6`](https://github.com/nodejs/io.js/commit/0b122d09c6)] - ***Revert*** "**dns**: remove AI_V4MAPPED hint flag on FreeBSD" (cjihrig) [iojs/io.js#1555](https://github.com/iojs/io.js/pull/1555)
* [[`42ed1e96ef`](https://github.com/nodejs/io.js/commit/42ed1e96ef)] - **doc**: update v8 flags in man page (Michaël Zasso) [iojs/io.js#1701](https://github.com/iojs/io.js/pull/1701)
* [[`d7ce2566d2`](https://github.com/nodejs/io.js/commit/d7ce2566d2)] - **deps**: update v8 to 4.3.61.21 (Chris Dickinson) [iojs/io.js#1632](https://github.com/iojs/io.js/pull/1632)
* [[`aa206c4d36`](https://github.com/nodejs/io.js/commit/aa206c4d36)] - **test**: remove obsolete harmony flags (Chris Dickinson)

yosuke-furukawa commented:
I would like to know the relationship between `next` and `next+1` branch.
io.js v3.0 is based on `next` branch or `next+1` branch ??

rvagg commented:
io.js v3.0.0 will be based on what's in this PR (may not come from this PR but it's this list of changes on `master`), `next` is where this comes from but `next` is falling behind `master` at a rapid rate, hence the change of approach. This means that v3.0.0 will be with V8 v4.4 unless we can't get it done for [some reason](https://github.com/nodejs/nan/issues/376).

`next+1` is where some folks are working with V8 v4.5, at the current rate I'd expect that work to form the basis of Node.js v4.0.0. Unfortunately `next+1` is going to suffer from the same problem as `next` in falling behind `master`. But if you want a playground for V8 v4.5 then you should go there.

rvagg commented:
v3.0.0 RC 4 based on this branch: https://iojs.org/download/rc/v3.0.0-rc.4/

```
$ node -pe process.versions
{ http_parser: '2.5.0',
  node: '3.0.0-rc.4',
  v8: '4.4.63.12',
  uv: '1.6.1',
  zlib: '1.2.8',
  ares: '1.10.1-DEV',
  modules: '45',
  openssl: '1.0.2d' }
```

Includes the node-gyp changes for quick header downloads (note how it not only grabs the headers tarball but it also knows this is an RC so it has to go elsewhere for it):

```
$ npm i buffertools
npm http request GET https://registry.npmjs.org/buffertools
npm http 200 https://registry.npmjs.org/buffertools
npm http fetch GET https://registry.npmjs.org/buffertools/-/buffertools-2.1.3.tgz
npm http fetch 200 https://registry.npmjs.org/buffertools/-/buffertools-2.1.3.tgz
npm WARN engine buffertools@2.1.3: wanted: {"node":">=0.3.0"} (current: {"node":"3.0.0-rc.4","npm":"2.13.1"})

> buffertools@2.1.3 install /home/rvagg/git/iojs/io.js/node_modules/buffertools
> node-gyp rebuild

gyp http GET https://iojs.org/download/rc/v3.0.0-rc.4/iojs-v3.0.0-rc.4-headers.tar.gz
gyp http 200 https://iojs.org/download/rc/v3.0.0-rc.4/iojs-v3.0.0-rc.4-headers.tar.gz
gyp http GET https://iojs.org/download/rc/v3.0.0-rc.4/SHASUMS256.txt
gyp http 200 https://iojs.org/download/rc/v3.0.0-rc.4/SHASUMS256.txt
make: Entering directory '/home/rvagg/git/iojs/io.js/node_modules/buffertools/build'
  CXX(target) Release/obj.target/buffertools/buffertools.o
  SOLINK_MODULE(target) Release/obj.target/buffertools.node
  COPY Release/buffertools.node
make: Leaving directory '/home/rvagg/git/iojs/io.js/node_modules/buffertools/build'
buffertools@2.1.3 node_modules/buffertools
```

It also includes the `process.release` changes which will be used for further modifications to node-gyp which _should_ be upstreamed if all goes well.

```
$ node -pe process.release
{ name: 'io.js',
  sourceUrl: 'https://iojs.org/download/rc/v3.0.0-rc.4/iojs-v3.0.0-rc.4.tar.gz',
  headersUrl: 'https://iojs.org/download/rc/v3.0.0-rc.4/iojs-v3.0.0-rc.4-headers.tar.gz' }
```

ralphtheninja commented:
Huge +1 for `process.release`

ofrobots commented:
@rvagg I'm a little bit confused with what the difference between the `next` and this branch is and why it would be easier to merge `master` onto this branch and then cherry-pick from `next` rather than merging `master` onto `next`.

Going forward, should V8 upgrades (like the latest one in #2220) continue going into `next`?

trevnorris commented:
@rvagg Buffer changes look good to me. Though I would recommend we land https://github.com/nodejs/io.js/pull/2195 first. Also I noticed another bug in `Slice()`. Easy enough to fix, but will need to submit a PR and all that.

trevnorris commented:
@rvagg Here's the fix for the other issue: https://github.com/nodejs/io.js/pull/2225

rvagg commented:
@ofrobots this:

```
$ git diff v3.0.0-alpha next --stat | tail -1
 695 files changed, 13678 insertions(+), 18437 deletions(-)
```

Granted a lot of that is npm, but getting out release candidates from a `next` branch that is so far behind `master` is a pain and lacks legitimacy. This is just my attempt at resetting so we have a path to getting it out.

I was rightly scolded in the TSC call today for my willingness to rebase given the pain this causes for anyone else working on the same branch, but we're starting to converge on a better process for keeping all of these things up to date. 3.0.0 is just a hump we have to get over now!

ofrobots commented:
Okay, I see. Going forward I can volunteer some time to try and keep `next` and `next+1` up-to-date with master.

hokein commented:
Hi all, could you tell me when will iojs v3.0.0 will release?

Fishrock123 commented:
> Hi all, could you tell me when will iojs v3.0.0 release?

Soon™ -- Literally when it's ready. Most of everything is done now, we had to wait for NAN to catch up for native addons to be reasonable.

bnoordhuis commented:
> Going forward I can volunteer some time to try and keep next and next+1 up-to-date with master.

Me too post.  I've been a little overcommitted recently but I'll have more time again starting next week.

I missed the discussion on yesterday's TSC call.  Is the plan to merge master frequently into next and next+1 (or master into next and next into next+1)?

Fishrock123 commented:
@bnoordhuis will you possibly be able to make it to #2230?

The generally discussion has shifted a bit. Turns out things like bisect and blame don't work so well with the previous proposed approach.

See also: https://gist.github.com/trevnorris/7620a64b086e95271197

Fishrock123 commented:
@bnoordhuis will you possibly be able to make it to #2230?

The generally discussion has shifted a bit. Turns out things like bisect and blame don't work so well with the previous proposed approach.

See also: https://gist.github.com/trevnorris/7620a64b086e95271197 and https://github.com/nodejs/io.js/issues/2215

bnoordhuis commented:
> will you possibly be able to make it to #2230?

I think I can make that.  I'll try to.

rvagg commented:
https://iojs.org/download/rc/v3.0.0-rc.5/ RC 5

Starefossen commented:
Great work :clap:

YuriSolovyov commented:
https://github.com/nodejs/io.js/pull/2193 should probably included too, according to  https://github.com/nodejs/io.js/pull/2193#issuecomment-124946048

Fishrock123 commented:
@YuriSolovyov everything in 2.x will be included. Don't worry. :)

trevnorris commented:
There have been additional bug fixes (e.g. 9cf36a1be). Will these be taken into the v3.0.0 release, or done after in a 3.0.1?

AntouanK commented:
Can we expect Chrome's ES6 features in v3?
( like arrow functions, spread operator and Array static methods )
see http://kangax.github.io/compat-table/es6/#chrome45

targos commented:
@AntouanK probably not, because Chrome 45 is still in beta and v3 will be released with the V8 version of Chrome 44.

bnoordhuis commented:
@AntouanK v3.x will ship with V8 4.4, the same version of V8 that's in Chrome 44.

rvagg commented:
Updated to latest master, included additional items in next that were not in here, performed a diff against a next + master merged and found some rebase problems that I've fixed, also discovered that the commit for https://github.com/nodejs/io.js/pull/1961 hasn't been pulled in to next yet.

Merged in #2275 and #2247 prior to cutting a v3.0.0-rc.6 which is now available in https://iojs.org/download/rc/v3.0.0-rc.6/ - @fhemberger would you mind testing both Windows installers? I haven't tried them but they appear to have been generated successfully.

rvagg commented:
need a bit more testing and to have some of the commits I've included here properly signed off and merged (those two I mentioned in particular) but the next step is to move this to a release proposal for v3.0.0. I believe this is basically ready.

AntouanK commented:
@targos @bnoordhuis Thanks. So we'll wait for v8 4.5? And that will go in next io/node major, like v4.x?

fhemberger commented:
@rvagg Tested iojs-v3.0.0-rc.6-x64.msi on Windows 8.1 and iojs-v3.0.0-rc.6-x86.msi on Windows 7 – both are working fine. However, on the installation confirmation screen, one placeholder is not replaced properly and the message reads "[ProductName] has been successfully installed". I'm looking into it and will update my PR.

fhemberger commented:
@rvagg In commit a0e994bb7befa0008c870d79f1658ef460849fcf, I removed the "[ProductName]" reference from the string, so we should be good.

rvagg commented:
@fhemberger cool, squashed that commit with your original PR commit and it's at the HEAD of this branch atm, still waiting for approval

rvagg commented:
Can someone please do a V8 upgrade to `next` for 4.4.65?

## 4.4.65 (2015-05-13)

  * Deprecate Isolate::New.
  * Factor out core of Array.forEach and .every, for use in TypedArrays
    (issue 3578).

rvagg commented:
Trying to do a changelog, focusing on V8 changes for now. The changelog for 4.3 and 4.4 are here: https://gist.github.com/rvagg/1f115074cb3c890985bf

This is what I have now:

- Introduction of `Maybe` and `MaybeLocal` C++ API for objects which _may_ or _may not_ have a value.
- `Array` can now be subclassed
- Removed `--harmony-scoping` flag
- Removed `--harmony-numeric-literal` flag
- Removed `SetIndexedPropertiesToExternalArrayData` and related APIs, forcing a shift to `Buffer` to be reimplemented based on `Uint8Array`

I'm just not sure if we've added any ES6 features here, we got numeric literals in 2.x didn't we? and scoping has been done for a while too right?

@domenic can you help here?

Fishrock123 commented:
> `Array` can now be subclassed

Technically this is an ES6 feature. Not syntax sugar to get excited about, but very important technically.

targos commented:
@rvagg AFAIK the stable V8 4.4 branch is 4.4.63 (it is the one that gets patches and that Chrome is using)

Also I think it would be good to mention changes about ES6 features in their own category as it is probably one of the things that the users are the most interested about.
Some of the additions: 
* Computed properties in object literals
* Unicode code point escapes

rvagg commented:
@targos why am I not seeing these in the V8 changelog? The things I listed are what I gleaned from there.

targos commented:
I guess the changelog is not very complete.
You can get a lot more [using git](https://gist.github.com/targos/d21095f6138ba5d6a185), but then it is complex work to read andunderstand what changes are relevant...

domenic commented:
> Unicode code point escapes

These are not listed on chromestatus.com, and last I tested were very buggy (even in Canary). Can you confirm they are actually working?

domenic commented:
Branch head is 4.4.63.26 BTW

rvagg commented:
@nodejs/collaborators @nodejs/tsc I've renamed this PR to "Release proposal: v3.0.0" and have prepared an initial version of the CHANGELOG.md, view it in full here: https://github.com/nodejs/io.js/blob/v3.0.0-alpha/CHANGELOG.md

Notable items copied here because I need help with (a) getting them right and (b) making sure I have everything listed that should be (thanks for the V8 chagelog @targos but it's a train wreck as usual so I'm still stabbing in the dark here):

### Notable changes

* **buffer**:
  - Due to changes in V8, it has been necessary to reimplement `Buffer` on top of V8's `Uint8Array`. While every effort has been made to maintain performance, users are likely to experience a different performance profile depending on how `Buffer` is used. (Trevor Norris) [#1825](https://github.com/nodejs/io.js/pull/1825).
  - `Buffer` can now take `ArrayBuffer`s as a constructor argument (Trevor Norris) [#2002](https://github.com/nodejs/io.js/pull/2002).
  - When a single buffer is passed to `Buffer.concat()`, a new, copied `Buffer` object will be returned; previous behavior was to return the original `Buffer` object (Sakthipriyan Vairamani) [#1937](https://github.com/nodejs/io.js/pull/1937).
* **build**: PPC support has been added to core to allow compiling on pLinux BE and LE (AIX support coming soon) (Michael Dawson) [#2124](https://github.com/nodejs/io.js/pull/2124).
* **dgram**: If an error occurs within `socket.send()` and a callback has been provided, the error is only passed as the first argument to the callback and not emitted on the `socket` object; previous behavior was to do both (Matteo Collina & Chris Dickinson) [#1796](https://github.com/nodejs/io.js/pull/1796)
* **freelist**: Deprecate the undocumented `freelist` core module (Sakthipriyan Vairamani) [#2176](https://github.com/nodejs/io.js/pull/2176).
* **http**: Status codes now all use the official [IANA names](http://www.iana.org/assignments/http-status-codes) as per [RFC7231](https://tools.ietf.org/html/rfc7231), e.g. `http.STATUS_CODES[414]` now returns `'URI Too Long'` rather than `'Request-URI Too Large'` (jomo) [#1470](https://github.com/nodejs/io.js/pull/1470).
* **node**:
  - `NODE_MODULE_VERSION` has been bumped to `45` to reflect the break in ABI (Rod Vagg) [#2096](https://github.com/nodejs/io.js/pull/2096).
  - Introduce a new `process.release` object that contains a `name` property set to `'io.js'` and `sourceUrl`, `headersUrl` and `libUrl` (Windows only) properties containing URLs for the relevant resources; this is intended to be used by node-gyp (Rod Vagg) [#2154](https://github.com/nodejs/io.js/pull/2154).
  - The version of node-gyp bundled with io.js now downloads and uses a tarball of header files from iojs.org rather than the full source for compiling native add-ons; it is hoped this is a temporary floating patch and the change will be upstreamed to node-gyp soon (Rod Vagg) [#2066](https://github.com/nodejs/io.js/pull/2066).
* **smalloc**: The `smalloc` module has been removed as it is no longer possible to provide the API due to changes in V8 (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022).
* **tls**: Add `server.getTicketKeys()` and `server.setTicketKeys()` methods for decryption of the [TLS session keys](https://www.ietf.org/rfc/rfc5077.txt) (Fedor Indutny) [#2227](https://github.com/nodejs/io.js/pull/2227).
* **v8**: Upgraded to 4.4.63.9 (Ben Noordhuis) [#2022](https://github.com/nodejs/io.js/pull/2022)
  - ES6: Enabled [computed property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)
  - ES6: `Array` can now be subclassed
  - ES6: Implement rest parameters in staging, use the `--harmony-rest-parameters` command line flag
  - ES6: Implement the spread operator in staging, use the `--harmony-spreadcalls` command line flag
  - Removed `SetIndexedPropertiesToExternalArrayData` and related APIs, forcing a shift to `Buffer` to be reimplemented based on `Uint8Array`
  - Introduction of `Maybe` and `MaybeLocal` C++ API for objects which _may_ or _may not_ have a value.
  - Added support for PPC

-------------------------

## Proposal

I propose that we do the following:

* Pull in final commits, currently as far as I'm aware the only thing we should prioritise is https://github.com/nodejs/io.js/pull/2220 if possible to get V8 up to 4.4.63.25, _although_ there is a 4.4.63.26 now which would be even better to get to.
* Land it to `master` and cut the release off `master`
* Create a new branch called `v3.x` for further v3.x.y releases and `master` will then become our "unstable" branch (much like `next` is now), this is (I believe) in line with the new release policy we've discussed to death, details found [here](https://github.com/nodejs/dev-policy/pull/75/files)
* Delete the `next` branch

Additionally (although I don't want bikeshedding on these to hold up release):

* `master` can land semver-major patches and new V8 upgrades but only once those versions of V8 become stable. Currently 4.5 is sitting in `next+1` but it's only beta so it shouldn't be landed yet.
* `master` will be the new base for Node.js v4, therefore, we can start migrating all of our active work into the https://github.com/nodejs/node repo, including the `v3.x` branch which will continue to be used for io.js releases (I don't see a good reason to keep this repo active except for all of the active PRs and issues). I'd also be happy for us to move v0.10 and v0.12 branches there too so all of our releases come from the same repo.

indutny commented:
I'd say those tls APIs are for ticket key rotation, not decryption. Technically you are right, but the use case is a bit off. Thanks!

thefourtheye commented:
Can we remove `sys` and `freelist` modules as well in 3.0.0?

rvagg commented:
@thefourtheye bit late for that discussion, perhaps open a new issue to discuss it for 4.0, freelist is _deprecated_ here, it hasn't been deprecated before so removal would likely be against our previous policies on the matter

thefourtheye commented:
@rvagg Oh okay. I created an issue as you suggested #2285 2285

rvagg commented:
https://jenkins-iojs.nodesource.com/job/node-test-commit/70/ looking good

@joaocgreis & @orangemocha we have some git problems on [15.04](https://jenkins-iojs.nodesource.com/job/node-test-commit-linux/62/nodes=ubuntu1504-64/) and [fedora22](https://jenkins-iojs.nodesource.com/job/node-test-commit-linux/62/nodes=fedora22/console)

orangemocha commented:
> @joaocgreis & @orangemocha we have some git problems on 15.04 and fedora22

Yes, they are the same that @joaocgreis already tracked down on Ubuntu 12.04. We need an updated git.

thefourtheye commented:
@orangemocha Is there an ETA to update the git? I am holding off few PRs, because of that machine's unavailability.

orangemocha commented:
@thefourtheye Monday at the latest.

thefourtheye commented:
@orangemocha Cool. Thanks for the update :-)

silverwind commented:
https://github.com/nodejs/io.js/pull/2224 should also make it. I'd propose a changelog entry like this once it has landed:

- **repl**: persistent history is now enabled by default. By default, the history file is located at `~/.node_repl_history`, which can be overridden by the new environment variable `NODE_REPL_HISTORY` which deprecates the previous `NODE_REPL_HISTORY_FILE`. Additionally, the format of the file has changed to plain text to better handle file corruption. To migrate to the new file and format, set the new variable if you need a custom location, run the REPL once, unset the old variable and delete the old history file.

@Fishrock123 does that sound good?

Fishrock123 commented:
I can't test #2224 now because the CI is dead. AGAIN. (see: https://github.com/nodejs/build/issues/147) (I still want to build a jenkins alternative...)

@silverwind some edits:

"which can be overridden by the new environment variable `NODE_REPL_HISTORY`. This deprecates the previous `NODE_REPL_HISTORY_FILE` variable"

"has been changed to plain text to better handle file corruption"

> To migrate to the new file and format, set the new variable if you need a custom location, run the REPL once, unset the old variable and delete the old history file.

Not sure about this bit. (is it needed? too much info?) We should note that setting `NODE_REPL_HISTORY=""` does disable the functionality.

YuriSolovyov commented:
> (I still want to make a jenkins alternative...)

Why not Travis?

Fishrock123 commented:
> Why not Travis?

Travis doesn't support running on multiple OS types and machines/architectures AFAIK.

silverwind commented:
> Not sure about this bit. (is it needed? too much info?) We should note that setting NODE_REPL_HISTORY="" does disable the functionality.

Yeah it's probably too heavy for the changelog. How about adding a small step-by-step for the migration to the docs in #2224 and link to it?

Fishrock123 commented:
If anyone has a **windows** machine, I have some persistent repl history tests for #2224 in https://github.com/Fishrock123/io.js/tree/repl-history-tests -- would be appreciated if you could run `./iojs --expose-internals test/sequential/test-repl-persistent-history.js` after building the branch.

rvagg commented:
Added #1761 back on /cc @trott @brendanashworth @Fishrock123, it was reverted on master and the reversion was included here. I mistakenly included the revert on `next` and I believe @brendanashworth was going to take care of that; although `next` won't be needed much longer ..

Let's aim to get this out today, I'll get back on to this by midday pacific time to give folks a chance to catch up with this to make sure we have everything covered. Still looking for help with the notable changes notes.

rvagg commented:
Fresh rebase on `master` which includes the npm upgrade. Pulled in the V8 upgrade from #2220 and running a new CI job: https://jenkins-iojs.nodesource.com/job/node-test-commit/79/ - I'm a bit concerned by the CI failures for #2220 that _look_ unrelated but aren't showing up on the last CI job I ran for this branch.

targos commented:
@rvagg this branch === `v3.0.0-alpha` or `next` ?
Failures on the Raspberries and Win2008 are fixed but only on master. The other ones didn't happen during my run two days ago.

orangemocha commented:
As far as CI goes, it is usable now though [the lint job is failing](https://jenkins-iojs.nodesource.com/job/node-linter/86/console) due to issues on the linter Jenkins machine.

orangemocha commented:
> (I still want to make a jenkins alternative...)

Last time I looked into it, [TeamCity](https://www.jetbrains.com/teamcity/) seemed like the most promising alternative.

rvagg commented:
this == `v3.0.0-alpha` which includes the latest `master`, looking good so far @ https://jenkins-iojs.nodesource.com/job/node-test-commit/79/

Fishrock123 commented:
#2224 hasn't been rebased so there may be some issues in that regard. I'll be up and looking to sort it out momentarily.

Fishrock123 commented:
Add to known bugs once #2224 lands:

```
* Using multiple REPL instances in parallel may cause some {recoverable} REPL history corruption. [#1634](https://github.com/nodejs/io.js/issues/1634)
```

I have yet to test #2224, but I think now it should still work if the REPL corrupts it. (unless utf8 fails to read and crashes in fs? Is that a thing?) I'll try to look into this part next..

silverwind commented:
Noteable change for #2224, based on @Fishrock123's feedback above:

- **repl**: persistent history is now enabled by default. By default, the history file is located at `~/.node_repl_history`, which can be overridden by the new environment variable `NODE_REPL_HISTORY`. This deprecates the previous `NODE_REPL_HISTORY_FILE` variable. Additionally, the format of the file has been changed to plain text to better handle file corruption.

brendanashworth commented:
@rvagg a noteable changelog for https://github.com/nodejs/io.js/commit/06cc919a0cd41380a83e4ea699f1a9ea30881266 / https://github.com/nodejs/io.js/pull/1617:

- **http**: calling `.getName()` on an HTTP agent no longer returns a trailing colon. HTTPS agents will no longer return an extra colon near the middle of the string.

rvagg commented:
pulled in @Fishrock123's repl changes @ #2224, updated the changelog, I believe this is ready to roll

rvagg commented:
blocked by https://github.com/nodejs/io.js/pull/2224#issuecomment-127425509

Fishrock123 commented:
Ugh, sorry. The rebase was pretty complex and I think I forgot to run the tests locally before handing off to Rod.

Fishrock123 commented:
The commits over on #2224 have been fixed up and rebased onto master.

There's a CI run with them on 3.0.0 at https://jenkins-iojs.nodesource.com/job/node-test-commit/96/

jbergstroem commented:
Fyi, the linters are running green lights again.

rvagg commented:
thanks @Fishrock123 and @jbergstroem 

I'm going to merge this all in to master in a moment, except for the release/changelog commit and start a new PR and move forward with the release

thefourtheye commented:
@rvagg Awesome news :-) Gotta keep a champagne bottle ready ;-)

mzgol commented:
Is `Array` really subclassable? Kangax tests at http://kangax.github.io/compat-table/es6/ report 0/11 for `Array` subclassing in Chrome 44.

Fishrock123 commented:
> Is `Array` really subclassable?

Well, TypedArrays are for buffer in any case. :p

mzgol commented:
Ah, it doesn't pass because it works only in strict mode but that is not a
problem in io.js; maybe it's worth mentioning this in changelog that it
works only in strict mode. But classes don't work at all outside strict in
v8 currently so whatever.

rvagg commented:
thanks @mzgol, have amended the changelog

rvagg commented:
https://jenkins-iojs.nodesource.com/job/node-test-commit/102/
& https://jenkins-iojs.nodesource.com/job/node-test-commit/103/
